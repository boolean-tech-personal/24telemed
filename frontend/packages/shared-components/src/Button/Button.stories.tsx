import { ButtonProps, Button } from './Button';
import type { Meta, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof Button>;

const meta: Meta<typeof Button> = {
  component:Button,
};

export default meta;

export const Primary: Story = {
  args: {
    color: 'white',
    onClick: () => console.log('clicked'),
    label: 'Click me',
    hoverColor: 'lightblue',
    activeColor: 'darkblue',
    focusColor: 'lightblue',
    backgroundColor: 'blue',
  },
};
export const Secondary: Story = {
  args: {
    color: 'white',
    onClick: () => console.log('clicked'),
    label: 'Click me',
    hoverColor: 'lightgreen',
    activeColor: 'darkgreen',
    focusColor: 'lightgreen',
    backgroundColor: 'green',
  },
};
export const Tertiary: Story = {
  args: {
    color: 'white',
    onClick: () => console.log('clicked'),
    label: 'Click me',
    hoverColor: 'lightgrey',
    activeColor: 'darkgrey',
    focusColor: 'lightgrey',
    backgroundColor: 'grey',
  },
};
