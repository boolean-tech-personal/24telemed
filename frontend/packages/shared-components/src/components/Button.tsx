import styled from 'styled-components';

export interface ButtonProps {
  color: string;
  onClick: () => void;
  label: string;
  hoverColor: string;
  activeColor: string;
  focusColor: string;
  backgroundColor: string;
}

const StyledButton = styled.button<ButtonProps>`
  background-color: ${(props) => props.backgroundColor};
    color: ${(props) => props.color};
    border: none;
    padding: 10px 20px;
    border-radius: 5px;
    cursor: pointer;
    transition: all 0.3s ease;
    &:hover {
      background-color: ${(props) => props.hoverColor };
    }
    &:active {
      background-color: ${(props) => props.activeColor };
    }
    &:focus {
      background-color: ${(props) => props.focusColor };
    }

    &:disabled {
      background-color: grey;
      cursor: not-allowed;
    }
    
`;

const Button = (props: ButtonProps) => {
  return (
    <StyledButton {...props} onClick={props.onClick}>
      {props.label}
    </StyledButton>
  );
};

export default Button;
