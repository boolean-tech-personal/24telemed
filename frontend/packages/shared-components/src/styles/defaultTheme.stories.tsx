import type { Meta } from "@storybook/react";
import { capitalize, get } from "lodash-es";
import styled, { css, useTheme } from "styled-components";

import type { PaletteVariants, Theme } from "./theme.types";

interface ColorProps {
  variant: PaletteVariants;
}

function Color({ variant }: ColorProps) {
  const theme = useTheme() as Theme;
  return (
    <ColorRoot variant={variant}>
      <div>
        {capitalize(variant.split(/(\d)./g).join(" ").split(".").join(" "))}
      </div>
      <div>{get(theme.palette, variant)}</div>
    </ColorRoot>
  );
}

const ColorRoot = styled.div<ColorProps>`
  width: 300px;
  height: 137px;
  border-radius: 10px;
  padding: 24px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  ${({ theme }) => css`
    border: 2px solid ${theme.palette.neutral.main};
  `}

  ${({ theme, variant }) => css`
    color: ${get(theme.onPalette, variant)};
    background-color: ${get(theme.palette, variant)};
  `}
`;

function Layout() {
  return (
    <div>
      <Row>
        <Color variant="primary1.main" />
        <Color variant="primary1.light" />
        <Color variant="primary1.lighter" />
      </Row>
      <Row>
        <Color variant="primary2.main" />
        <Color variant="primary2.light" />
        <Color variant="primary2.lighter" />
      </Row>
      <Row>
        <Color variant="secondary1.main" />
        <Color variant="secondary1.light" />
        <Color variant="secondary1.lighter" />
      </Row>
      <Row>
        <Color variant="secondary2.main" />
        <Color variant="secondary2.light" />
        <Color variant="secondary2.lighter" />
      </Row>
      <Row>
        <Color variant="neutral.main" />
        <Color variant="neutral.light" />
        <Color variant="neutral.lighter" />
      </Row>
      <Row>
        <Color variant="common.black" />
        <Color variant="common.white" />
      </Row>
      <Row>
        <Color variant="error" />
        <Color variant="success" />
        <Color variant="notification" />
      </Row>
    </div>
  );
}

const Row = styled.div`
  display: flex;
  gap: 24px;

  &:not(:last-child) {
    margin-bottom: 24px;
  }
`;

const Story: Meta<typeof Layout> = {
  component: Layout,
  title: "Components/Colors",
};
export default Story;

export const Colors = {
  args: {},
};
