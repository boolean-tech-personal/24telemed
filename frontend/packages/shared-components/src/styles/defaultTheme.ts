import type { Breakpoints, PaletteVariants, Theme } from "./theme.types";
import { createThemeBreakpoints } from "./utils";

export const breakpoints: Breakpoints = {
  xs: 600,
  sm: 900,
  md: 1200,
  lg: 1400,
};

const fontFamily = '"Uniform Rounded","Helvetica","Arial",sans-serif';
const fontFamilyHeading =
  '"Uniform Rounded Ex Cond","Helvetica","Arial",sans-serif';

export const colorVariants: Array<PaletteVariants> = [
  "primary1.main",
  "primary1.light",
  "primary1.lighter",
  "primary2.main",
  "primary2.light",
  "primary2.lighter",
  "secondary1.main",
  "secondary1.light",
  "secondary1.lighter",
  "secondary2.main",
  "secondary2.light",
  "secondary2.lighter",
  "secondary3.main",
  "secondary3.light",
  "secondary3.lighter",
  "neutral.main",
  "neutral.light",
  "neutral.lighter",
  "error",
  "success",
  "notification",
  "common.black",
  "common.white",
];

export const defaultTheme: Theme = {
  palette: {
    primary1: {
      main: "#2F4D52",
      light: "#829497",
      lighter: "#D5DBDC",
    },
    primary2: {
      main: "#FFF875",
      light: "#FFFBAC",
      lighter: "#FFFEE3",
    },
    secondary1: {
      main: "#5CD6C2",
      light: "#9DE6DA",
      lighter: "#DEF7F3",
    },
    secondary2: {
      main: "#ED7F75",
      light: "#F19991",
      lighter: "#FBE5E3",
    },
    secondary3: {
      main: "#FFB06C",
      light: "#FFD0A7",
      lighter: "#FFEFE2",
    },
    neutral: {
      main: "#EAEDEE",
      light: "#F2F4F5",
      lighter: "#FBFBFC",
    },
    common: {
      black: "#000000",
      white: "#FFFFFF",
      transparent: "transparent",
    },
    background: {
      main: "#F7F8F8",
      secondary: "#EFF5F6",
      white: "#FFFFFF",
    },
    error: "#FF6363",
    success: "#6FCF97",
    notification: "#9DE6DA",
  },
  onPalette: {
    primary1: {
      main: "#FFFFFF",
      light: "#FFFFFF",
      lighter: "#2F4D52",
    },
    primary2: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    secondary1: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    secondary2: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    secondary3: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    neutral: {
      main: "#2F4D52",
      light: "#2F4D52",
      lighter: "#2F4D52",
    },
    common: {
      black: "#FFFFFF",
      white: "#2F4D52",
      transparent: "#2F4D52",
    },
    background: {
      main: "#2F4D52",
      secondary: "#2F4D52",
      white: "#2F4D52",
    },
    error: "#FFFFFF",
    success: "#FFFFFF",
    notification: "#2F4D52",
  },
  breakpoints: createThemeBreakpoints(breakpoints),
  typography: {
    htmlFontSize: 16,
    fontFamily: fontFamily,
    fontWeight: {
      regular: 400,
      bold: 700,
    },
    defaultColor: "#2F4D52",
    h1: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "3.5rem",
      fontSizeSm: "3rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h2: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "3rem",
      fontSizeSm: "2.25rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h3: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "2.25rem",
      fontSizeSm: "2rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h4: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "2rem",
      fontSizeSm: "1.75rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h5: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "1.75rem",
      fontSizeSm: "1.5rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    h6: {
      fontFamily: fontFamilyHeading,
      fontWeight: 700,
      fontSize: "1.5rem",
      fontSizeSm: "1.25rem",
      lineHeight: 0.96,
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    bodyLg: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "1.125rem",
      lineHeight: 1.5,
    },
    bodyMd: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "1rem",
      lineHeight: 1.5,
    },
    bodySm: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "0.875rem",
      lineHeight: 1.5,
    },
    bodyXs: {
      fontFamily: fontFamily,
      fontWeight: 400,
      fontSize: "0.75rem",
      lineHeight: 1.5,
    },
    button: {
      fontFamily: fontFamily,
      fontWeight: 700,
      fontSize: "1rem",
      lineHeight: 1.5,
      height: "3.5rem",
      letterSpacing: "-0.02em",
      textTransform: "uppercase",
    },
    navigationButton: {
      fontFamily: fontFamily,
      fontWeight: 700,
      fontSize: "1rem",
      lineHeight: 1.5,
      letterSpacing: "-0.02em",
      height: "2.5rem",
      textTransform: "uppercase",
    },
    link: {
      fontWeight: 700,
      textDecoration: "underline",
      fontSize: "1.125rem",
      fontFamily: fontFamily,
    },
  },
  radius: {
    none: "0px",
    progressBar: "4px",
    languageSelect: "8px",
    notification: "8px",
    dropdown: "10px",
    tooltip: "10px",
    textarea: "10px",
    searchResultCard: "10px",
    card: "20px",
    button: "99px",
    toast: "999px",
  },
  border: {
    primary: {
      main: "2px solid #2F4D52",
      light: "1px solid rgba(47, 77, 82, 0.20)",
    },
    secondary: {
      main: "2px dashed #2F4D52",
      light: "2px dashed rgba(47, 77, 82, 0.20)",
    },
    bold: {
      main: "6px solid #2F4D52",
      light: "5px solid #2F4D52",
    },
  },
  spacing: {
    none: "0px",
    xxs: "4px",
    xs: "8px",
    sm: "16px",
    md: "24px",
    lg: "32px",
    xl: "48px",
  },
  height: {
    navBar: "80px",
  },
  sizes: ["xs", "sm", "md", "lg", "xl"],
  shadows: {
    // TODO: discuss also about the shadows, it inherently convey a meaning of elevation depending on how far the shadow is projected
    // I don't think with the current use we have it's working well
    none: "none",
    input: "0px 0px 10px 0px rgba(47,77,82,0.25)",
    md: "8px 13px 30px 0px #2F4D521A",
    lg: "3px 4px 15px 0px #2F4D5240",
    xl: "8px 13px 30px 0px #2F4D5240",
    inset: "3px 4px 10px 0px #2F4D5240 inset",
  },
  icon: {
    fontSizes: {
      sm: "1rem",
      md: "1.5rem",
      lg: "3rem",
    },
  },
  opacity: {
    disabled: 0.3,
    muted: 0.6,
    dimmed: 0.8,
  },
  transitions: {
    easing: {
      easeInOut: "cubic-bezier(0.4, 0, 0.2, 1)",
      easeOut: "cubic-bezier(0.0, 0, 0.2, 1)",
    },
    duration: {
      shortest: 160,
      shorter: 175,
      short: 250,
      standard: 300,
      complex: 375,
      enteringScreen: 225,
      leavingScreen: 195,
    },
  },
  zIndex: {
    appBar: 1200,
    modal: 1500,
    toaster: 1600,
    selectPopper: 1700,
    tooltip: 1800,
  },
  button: {
    primary: {
      backgroundColor: "#2F4D52",
      color: "#FFFFFF",
      hoverBackgroundColor: "#D5DBDC",
      disabledBackgroundColor: "#D5DBDC",
      disabledColor: "#829497",
      hoverColor: "#FFFFFF",
      focusBackgroundColor: "#2F4D52",
      pressedBackgroundColor: "#829497",
    },
    secondary: {
      backgroundColor: "#EAEDEE",
      color: "#2F4D52",
      hoverBackgroundColor: "#EFF5F6",
      disabledBackgroundColor: "#D5DBDC",
      disabledColor: "#829497",
      hoverColor: "##FFFFFF",
      focusBackgroundColor: "#EAEDEE",
      pressedBackgroundColor: "#829497",
    },
    tertiary: {
      backgroundColor: "transparent",
      color: "#2F4D52",
      hoverBackgroundColor: "#EFF5F6",
      disabledBackgroundColor: "transparent",
      disabledColor: "#829497",
      hoverColor: "#2F4D52",
      focusBackgroundColor: "transparent",
      pressedBackgroundColor: "#D5DBDC",
    },
  },
};
