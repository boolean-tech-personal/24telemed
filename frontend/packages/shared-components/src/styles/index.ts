export { defaultTheme, colorVariants, breakpoints } from "./defaultTheme";
export * from "./theme.types";
export * from "./utils";
export * from "./createTheme";
export * from "./useBreakpoints";
