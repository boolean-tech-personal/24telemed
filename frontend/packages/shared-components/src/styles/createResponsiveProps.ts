import { css } from "styled-components";

import { breakpoints } from "./defaultTheme";
import { Breakpoints, Theme } from "./theme.types";

export type ResponsivePropsBase = Record<string, unknown>;

export type ResponsiveProps<TResponsivePropsBase extends ResponsivePropsBase> =
  {
    [K in keyof TResponsivePropsBase as `${keyof Breakpoints}${Capitalize<
      string & K
    >}`]?: TResponsivePropsBase[K];
  };

export type PropsCssMap<TResponsivePropsBase extends ResponsivePropsBase> =
  Required<{
    [K in keyof TResponsivePropsBase as Capitalize<string & K>]: string;
  }>;

export type ResponsivePropsGetters<
  TResponsivePropsBase extends ResponsivePropsBase,
> = Partial<{
  [K in keyof PropsCssMap<TResponsivePropsBase>]: (
    theme: Theme,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    value: any,
  ) => string;
}>;

export const createResponsiveProps = <
  TResponsivePropsBase extends ResponsivePropsBase,
>(
  propsCssMap: PropsCssMap<TResponsivePropsBase>,
  responsivePropsGetters?: ResponsivePropsGetters<TResponsivePropsBase>,
) =>
  function responsiveProps(
    props: ResponsiveProps<TResponsivePropsBase> & { theme: Theme },
  ) {
    const breakpointKeys = Object.keys(breakpoints) as Array<keyof Breakpoints>;
    const cssMapEntries = Object.entries(propsCssMap);

    const styles = [];
    for (let i = 0; i < breakpointKeys.length; i++) {
      const breakpoint = breakpointKeys[i];
      let cssProperties = "";

      for (let j = 0; j < cssMapEntries.length; j++) {
        const [propBase, cssProp] = cssMapEntries[j] as [
          keyof PropsCssMap<TResponsivePropsBase>,
          string,
        ];
        if (typeof propBase !== "string") continue;

        const prop =
          `${breakpoint}${propBase}` as keyof ResponsiveProps<TResponsivePropsBase>;
        const value = props[prop];

        if (value !== undefined) {
          const transformedValue = responsivePropsGetters?.[propBase]?.(
            props.theme,
            value,
          );
          cssProperties += `${cssProp}: ${transformedValue ?? value};`;
        }
      }

      if (cssProperties) {
        styles.push(
          props.theme.breakpoints[breakpoint].down(css`
            ${cssProperties}
          `),
        );
      }
    }

    return styles;
  };
