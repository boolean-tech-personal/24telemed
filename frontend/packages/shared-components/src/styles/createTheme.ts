import defaultsDeep from "lodash/defaultsDeep";
import type { PartialDeep } from "type-fest";

import { defaultTheme } from "./defaultTheme";
import type { Theme } from "./theme.types";

export function createTheme(themeOverride: PartialDeep<Theme> = {}): Theme {
  return defaultsDeep(themeOverride, defaultTheme);
}
