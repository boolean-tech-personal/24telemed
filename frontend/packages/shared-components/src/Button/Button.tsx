import styled, {css} from 'styled-components';
import { FC } from 'react';
import { makeButtonVariant } from '../styles';
import { ResponsivePropsBase } from '../styles/createResponsiveProps';
import type { NewButtonVariant } from '../styles';

export interface BaseButtonProps extends ResponsivePropsBase {
  text: string;
  variant?: NewButtonVariant;
  rightIcon?: React.ReactNode;
  disabled?: boolean;
  leftIcon?: React.ReactNode;
}

const StyledButton: FC<BaseButtonProps> = styled.button.attrs<BaseButtonProps>(
  ({ variant = 'primary', disabled, rightIcon, leftIcon, text }) => ({
    variant,
    disabled,
    rightIcon,
    leftIcon,
  })
)`
  ${makeButtonVariant}
  width: 100%;
  padding-top: ${({ theme }) => ` ${theme.spacing.sm}`};
  padding-bottom: ${({ theme }) => ` ${theme.spacing.sm}`};
  padding-left: ${({ theme, leftIcon }) =>
    ` ${leftIcon ? theme.spacing.xl : theme.spacing.sm}`};
  padding-right: ${({ theme, rightIcon }) =>
    ` ${rightIcon ? theme.spacing.xl : theme.spacing.sm}`};
  border-radius: 12px;
  border: none;
  cursor: pointer;
  font-family: ${({ theme }) => theme.typography.ontFamily};
  border: ${({ theme }) => theme.border.primary.light};
  text-align: center;
  border-color: ${({ theme, variant }) => variant === 'tertiary' ? theme.border.primary.light : "transparent"};
  ${({ theme }) =>
    theme.breakpoints.xs.down(css`
      font-size: ${({ theme }) => theme.typography.bodyMd.fontSize};
      height: ${({ theme }) => theme.spacing.md};
      display: flex;
      align-items: center;
      justify-content: center;
    `)}
`;

export function Button(props: BaseButtonProps) {
  return (
    <ButtonContainer>
      <StyledButton
        {...props}
        disabled={props.disabled}
        rightIcon={props.rightIcon}
        variant= {props.variant}
      >
        {props.text}
      </StyledButton>
      {props.rightIcon && (
        <RightIconContainer>{props.rightIcon}</RightIconContainer>
      )}

      {props.leftIcon && (
        <LeftIconContainer>{props.leftIcon}</LeftIconContainer>
      )}
    </ButtonContainer>
  );
}

const ButtonContainer = styled.div`
  position: relative;
  width: 100%;
`;

const RightIconContainer = styled.div`
  position: absolute;
  right: ${({ theme }) => theme.spacing.sm};
  top: 50%;
  transform: translateY(-50%);
`;

const LeftIconContainer = styled.div`
  position: absolute;
  left: ${({ theme }) => theme.spacing.sm};
  top: 50%;
  transform: translateY(-50%);
`;
