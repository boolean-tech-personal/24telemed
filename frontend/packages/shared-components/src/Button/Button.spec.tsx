import '@testing-library/jest-dom/extend-expect';
import { Button } from './Button';
import { render, screen, fireEvent } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '../styles/defaultTheme';

const TestButton = () => {
  return (
    <ThemeProvider theme={defaultTheme}>
      <Button text="Click me" />
    </ThemeProvider>
  );
};

describe('Button', () => {
  it('should render the button', () => {
    render(<TestButton />);
    expect(screen.getByText('Click me')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<TestButton />);
    expect(container).toMatchSnapshot();
  });

    it('should call the onClick function', () => {
        // @ts-ignore
        const onClick = jest.fn();
        render(<Button text="Click me" onClick={onClick} />);
        fireEvent.click(screen.getByText('Click me'));
        expect(onClick).toHaveBeenCalled();
    });

});
