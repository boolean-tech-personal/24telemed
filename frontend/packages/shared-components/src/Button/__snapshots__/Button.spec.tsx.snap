// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-bcXHqe fLEOWP"
    color="white"
    label="Click me"
  >
    Click me
  </button>
</DocumentFragment>
`;
